### PR TITLE
Smoothing in gammapy maps

### PR DESCRIPTION
**Pull request by AtreyeeS**
Please delete PR #1430 and consider this one.

### CODE REVIEW

File: gammapy/maps/wcsnd.py [hunk 1]
```diff
@@ -2,8 +2,11 @@
 from __future__ import absolute_import, division, print_function, unicode_literals
 import copy
 import numpy as np
+import matplotlib.pyplot as plt
```
Comment (adonath) [inline gammapy/maps/wcsnd.py]: Matplotlib is actually an optional dependency, this import should go back to the `WcsNDMap.plot()` method.

**Comment by adonath** [comment]
Thanks, the file conflicts are resolved now! Still it seems like there was a copy & paste error. I guess you wanted to add the smoothing to existing `WcsNDMap.plot()` method and not add a new plot method. Can you please check your changes again?

**Comment by AtreyeeS** [comment]
Hello,

Sorry for the error!
I have now deleted the old plot function. Hope that is OK

regards,

**Comment by adonath** [comment]
Thanks, no need to apologize! As I've written in my comment above it would still help if you just added your changes to the existing function, instead of adding a new one and deleting the old one. Does that make sense?

**Comment by AtreyeeS** [comment]
Well, I changed quite a few small things
eg: using simple_norm from astropy visualization instead of the matplotlib colors, and added a "add_cbar" and "stretch" options.
So the entire function is changed apart from the slicing part:
        if idx is not None:
            idx = (idx,) if isinstance(idx, int) else idx
            slices = (slice(None), slice(None)) + idx
            data = self.data[slices[::-1]]
        else:
            data = self.data

So changing into the old routine would be changing most of the lines, do you really think that would be helpful?
I will try to change old functions properly in future.

**Comment by AtreyeeS** [comment]
Hello,

I think the problem was because of the position of the plot() that the diff was looking weird.
I have placed it properly now.
I think it should be ok. Please see.

**Comment by adonath** [comment]
@AtreyeeS Thanks, that's much better! I'll do a final review now, so that we can merge soon.

**Comment by adonath** [comment]
@AtreyeeS Sorry for the delay on this. I discussed with @cdeil offline and we both agreed that it might be better to add a separate `.smooth()` method to the `WcsNDMap` class, similar to [SkyImage.smooth()](http://docs.gammapy.org/dev/api/gammapy.image.SkyImage.html#gammapy.image.SkyImage.smooth). The smoothing should only apply to the spatial axes (either by looping over the other axes or simply requiring the map is only 2D) The method should then return again a `WcsNDMap` object, that can be used for plotting. Are you motivated to change this PR to implement a `WcsNDMap.smooth()` method instead? This work would involve to add the `.smooth()` method and docstring (large parts can be copied and pasted from `SkyImage.smooth()`) and a small test of the smoothing in `gammapy/maps/tests/test_wcsnd.py`.

**Comment by registerrier** [comment]
Hi @adonath and @cdeil, 
we discussed this also before and we convinced ourselves that creating a new ND map and smoothing it would be completely overkill for just plotting one slice of the map. As a user you want to see, your count map slice at 200 GeV smoothed at e.g. 0.15deg and your slice at 5 TeV smoothed at 0.05 deg. And you might want to change that it the result is not what you expect. 
A smoothing method on-the-fly is what is needed for plotting, IMO. 

This does not preclude the need for `Map.convolve()`method, but the latter would have different objectives.

**Comment by cdeil** [comment]
My 2 cents:

If it isn't already, we should make it very easy to get images from ND maps like:
```
map_2d = map_3d.get_by_coord({'energy': '1TeV'})
```
Maybe there's a missing utility method to make extracting image slices with our without copy easy?

Then `map.plot` and `map.smooth` could just compose nicely and one would do:
```
map.get_image({'energy': ' 200 GeV'}).smooth('0.15 deg').plot()
map.get_image({'energy': ' 5 TeV'}).smooth('0.05 deg').plot()
```
Or me personally, I often like to use DS9 to view images, so I would do this:
```
map.get_image({'energy': ' 5 TeV'}).smooth('0.05 deg').write('image.fits')
```
Note that this has some advantages:
- More modular and composable functionality, supporting more use cases
- Easier to test
- Easier to document and learn (fewer argument for each method)
- Plot doesn't need to grow fancy functionality to select images, that is super common, not just for plotting. If `get_image` is no-copy by default, it's just as efficient to chain with plot.

What if I asked you to add a test for the smooth functionality you are adding here, to establish the behaviour at the edge of the image, and that an image of constant value 42 comes out again, i.e. smoothing preserves flux / counts? It's super hard to test if you put compute code only in the plot method.

**Comment by adonath** [comment]
@cdeil This easy access of image planes is exactly what I started to work on yesterday evening (see https://github.com/gammapy/gammapy/compare/master...adonath:add_get_image_methods_to_maps). I could finish up this work and open a pull request today.

**Comment by cdeil** [comment]
@adonath - That sounds great, please open a PR when you have a proposal or any question. Also note #1329. It's about general map slicing, but the discussion is probably useful also when adding a method to get a single slice for now.

**Comment by cdeil** [comment]
So to be concrete, my suggestion for this PR would be:
- Move the smooth to a separate method. Add a test for each kernel and assert that a constant image is returned unchanged (assert on a corner pixel). Only support 2D map image for now, leave it up to the caller to make a 2D map before calling smooth.
- For plot, I would suggest to remove the `idx` option. Just assume in plot that that map is 2D as well, leaving it to the caller to slice out the image. This could be a separate PR.

Note that the `idx` option in plot already had to be augmented once (see #1384), and we'll have to augment it again and again to implement powerful slicing (see #1329). Note that the very similar and widely used and loved `xarray` also did slicing separately from plotting: http://xarray.pydata.org/en/stable/plotting.html

**Comment by registerrier** [comment]
OK why not. But then this means we need a specific object for 2D maps (images). This would be bad to have methods on `Map` that would expect only 2D instances. In xarray, as far as I understand, you have to project your DataSet to a specific object for plotting.

Then the simple/short term solution is to export to `SkyImage`, which has all the required functions. What we have to then is:
`SkyImage.from_nd_map(map.get_image({'energy': ' 200 GeV'})).smooth('0.15 deg').plot()`

Or the `WcsNDMap.get_image()`method has to return a `SkyImage`.

**Comment by cdeil** [comment]
> OK why not. But then this means we need a specific object for 2D maps (images). This would be bad to have methods on Map that would expect only 2D instances. 

I discussed this with @adonath yesterday, he made the same suggestion.
I don't think this is a good idea to add a subclass or wrapper class to special-case 2D.

There will be a handful of methods that only work for 2D, and for those we add these two lines at the top, to give a good error message:
```
if self.geom.ndim > 2:
    raise ValueError('Only supported on 2D maps')
```
This is e.g. also how scikit-image handles this, they have a mix of functions that work with 2D, 3D or ND.

> Then the simple/short term solution is to export to SkyImage

I added this a few months ago to help with the transition to Map:
* [SkyImage.to_wcs_nd_map](http://docs.gammapy.org/dev/api/gammapy.image.SkyImage.html#gammapy.image.SkyImage.to_wcs_nd_map)
* [SkyImage.from_wcs_nd_map](http://docs.gammapy.org/dev/api/gammapy.image.SkyImage.html#gammapy.image.SkyImage.from_wcs_nd_map)
and same methods on SkyCube.

You can use this for now, although if you want to contribute to Gammapy and help us move forward, the only sensible way is to contribute the ~ 100 lines of code that are still missing in `gammapy.maps` and to translate the existing code and docs using SkyImage and SkyCube to Map. This whole transition is ~ 2 weeks of work, and if we share it a bit, it can be done by the end of the coding sprint and we'll be in a much better place.

If this is controversial, maybe we can discuss it in the call tomorrow? (In that case, a clear second proposal would help, e.g. whether the special-cased maps classes would be subclasses or wrappers and what methods would go there and how it would interact with the ND maps methods)

**Comment by cdeil** [comment]
a-merge-conflict-i-sense

**Comment by cdeil** [comment]
This PR was superseded by #1444, which is already merged. Closing this one.
